src/java.base/share/classes/sun/security/pkcs12/PKCS12KeyStore.java

```diff
@@ -1434,7 +1435,7 @@ public synchronized void engineSetEntry(String alias, KeyStore.Entry entry,
     private Set<KeyStore.Entry.Attribute> getAttributes(Entry entry) {
 
         if (entry.attributes == null) {
-            entry.attributes = new HashSet<>();
+            entry.attributes = ConcurrentHashMap.newKeySet();
```

Comment:
It seems that there is still a risk that two threads may compete here, then you may end up with one thread working with a different copy of the attributes, no longer tied to the entry. E.g:

Thread#1 sees attributes is null, Thread#2 sees attributes is null, both set attributes and only one of them win. The caller in the second thread (that lost) is given the "dangling" map that has not been set, and if it modifies it then modifications will be lost.

Also I see in total four places that do:
```
[entry|this].attributes = new HashSet<>();
```

wouldn't you need to modify all four places in the same manner?

Reply:
I wonder if this would work, if you don't want to use final or volatile + CAS:

```
 class Entry { 
     ...
     private Set<KeyStore.Entry.Attribute> createAttributesMap() {
          synchronized (this) {
              var attributes = this.attributes;
              if (attributes == null) {
                  return attributes = this.attributes = ConcurrentHashMap.newKeySet();
              } else return attributes;
           }
       }
       
 }
 
 ...
 
      if (entry.attributes == null) {
          entry.attributes = entry.createAttributesMap();
      }
      
```